torch/_inductor/codegen/cpp_prefix.h

```diff
@@ -143,7 +149,141 @@ inline bool less_or_nan(scalar_t a, scalar_t b, int64_t idx_a, int64_t idx_b) {
   return (a == b) ? idx_a < idx_b : (a < b);
 }
 
+template <typename T>
+inline IndexValue<T>& argmin_combine(IndexValue<T>& a, T next_value, int64_t next_index){
+  if(!(less_or_nan(a.value, next_value, a.index, next_index))){
+    a.value = next_value;
+    a.index = next_index;
+  }
+  return a;
+}
+template <typename T>
+inline IndexValue<T>& argmax_combine(IndexValue<T>& a, T next_value, int64_t next_index){
+  if(!(greater_or_nan(a.value, next_value, a.index, next_index))){
+    a.value = next_value;
+    a.index = next_index;
+  }
+  return a;
+}
+template <typename T>
+inline IndexValue<T>& argmin_combine(IndexValue<T>& a, const IndexValue<T>& next){
+  if(!(less_or_nan(a.value, next.value, a.index, next.index))){
+    a.value = next.value;
+    a.index = next.index;
+  }
+  return a;
+}
+template <typename T>
+inline IndexValue<T>& argmax_combine(IndexValue<T>& a, const IndexValue<T>& next){
+  if(!(greater_or_nan(a.value, next.value, a.index, next.index))){
+    a.value = next.value;
+    a.index = next.index;
+  }
+  return a;
+}
+
 #if INDUCTOR_USE_VECTOR_TYPES()
+
+template <typename T, int L>
+struct IndexValueVec {
+  IndexValue<T> arr[L];
```

Comment:
Can we vectorize it instead of using arrays?

Reply:
`compare_op + blendv` should be worked if we do not have this issue here https://github.com/pytorch/pytorch/issues/43254#issuecomment-731288606. The aten's impl are also scalar version due to this issue.

Reply:
Can you elaborate more? Thanks.

Reply:
Hi, @jgong5. I re-read https://github.com/pytorch/pytorch/issues/43254 and I am wrong before. This issue is irelevant. It only impact when we use `void binary_kernel_reduce_vec(TensorIterator& iter, func_t op, vec_func_t vop, double ident = 0) {`, because of the `long int` loss precision while casting to `double ident.
I tried to vectorized argmin/argmax:
```
template <typename T, int NV, int NI>
inline IndexValueVec<T, NV, NI>& argmin_vec_impl(IndexValueVec<T, NV, NI>& a,  at::vec::VectorizedN<T, NV> value, at::vec::VectorizedN<int64_t, NI> index){
  at::vec::VecMask<T, NV> _mask(a.value > value);
  at::vec::VecMask<int64_t, NI> mask = _mask.template cast<int64_t, NI>();
  a.value = at::vec::minimum(a.value, value);
  a.index = at::vec::VectorizedN<int64_t, NI>::blendv(a.index, index, mask);
  return a;
}
``` 
It works fine if there is no `NaN`.
There is a special logical to handle `NaN` https://github.com/pytorch/pytorch/blob/b5b36cf0c4e1958f1ff25120f5d4beeef3288187/aten/src/ATen/native/SharedReduceOps.h#L419-L445
And I fail to vectorized this logic.

The the aten's impl for argmin/argmax is also scalar version
```
static void argmin_kernel_impl(TensorIterator &iter) {
  AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBFloat16, iter.dtype(1), "argmin_cpu", [&] {
    if (is_reduce_lastdim(iter)) {
      using arg_t = std::pair<scalar_t, int64_t>;
      auto op = ArgMinOps<scalar_t>{};
      binary_kernel_reduce_lastdim(iter, [&](char* result_data_bytes, char* self_data_bytes, int64_t size) {
        int64_t* result_data = (int64_t*)result_data_bytes;
        scalar_t* self_data = (scalar_t*)self_data_bytes;

        arg_t acc = arg_t(upper_bound<scalar_t>(), 0);
        for (int64_t i = 0; i < size; i++) {
          acc = op.reduce(acc, self_data[i], i);
        }
        result_data[0] = acc.second;
      });
      return;
    }
    binary_kernel_reduce(
      iter,
      ArgMinOps<scalar_t>{},
      std::pair<scalar_t, int64_t>(upper_bound<scalar_t>(), 0));
  });
}
```
Do you have any idea about vectorize the `NaN` logic?





`

Reply:
Can you use `isnan` to check?

Reply:
Vectorized now, and re-open some skipped ut since it is supported now